syntax = "proto2";

import "dotaservice/protos/dota_gcmessages_common_bot_script.proto";
import "dotaservice/protos/dota_shared_enums.proto";
// import "google/protobuf/empty.proto";

enum Status {
  // From google/rpc/code.proto
  OK = 0;
  RESOURCE_EXHAUSTED = 8;
  FAILED_PRECONDITION = 9;
}

enum HostMode {
  DEDICATED = 0;
  GUI = 1;
  GUI_MENU = 2;
}

message Empty {}

message Action {
  required CMsgBotWorldState.Action action = 1;
}

message Observation {
  required Status status = 1;
  optional CMsgBotWorldState world_state = 2;
}

message Config {
  optional string game_id = 1;
  required uint32 ticks_per_observation = 3;
  // TODO(tzaman): rework below, as host_mode impacts the settings (gui= timescale 1, etc)
  optional uint32 host_timescale = 2 [default = 1];
  optional HostMode host_mode = 5 [default = DEDICATED];
  optional DOTA_GameMode game_mode = 6 [default = DOTA_GAMEMODE_MO];
}

service DotaService {
  rpc reset (Config) returns (Observation) {}  // TODO(tzaman): add version and hostname in response.
  rpc step (Action) returns (Observation) {}
  rpc clear (Empty) returns (Empty) {}
}
